Cabal/Distribution/PackageDescription/Parsec.hs

```diff
@@ -0,0 +1,364 @@
+            message   = "This package requires at least Cabal version "
+                     ++ display cabalVersionNeeded
+    -}
+
+    checkForUndefinedFlags
+        :: GenericPackageDescription
+        -> ParseResult ()
+    checkForUndefinedFlags gpd = pure ()
+{-
+        let definedFlags = map flagName flags
+        mapM_ (checkCondTreeFlags definedFlags) (maybeToList mlib)
+        mapM_ (checkCondTreeFlags definedFlags . snd) sub_libs
+        mapM_ (checkCondTreeFlags definedFlags . snd) exes
+        mapM_ (checkCondTreeFlags definedFlags . snd) tests
+
+    checkCondTreeFlags :: [FlagName] -> CondTree ConfVar c a -> PM ()
+    checkCondTreeFlags definedFlags ct = do
+        let fv = nub $ freeVars ct
+        unless (all (`elem` definedFlags) fv) $
+            fail $ "These flags are used without having been defined: "
+                ++ intercalate ", " [ n | FlagName n <- fv \\ definedFlags ]
+-}
+
+parseConfVar :: [SectionArg Position] -> ParseResult ConfVar
+parseConfVar = error "parseConfVar"
+
+-- | Parse a list of fields, given a list of field descriptions,
+--   a structure to accumulate the parsed fields, and a function
+--   that can decide what to do with fields which don't match any
+--   of the field descriptions.
+parseFields :: [FieldDescr a]      -- ^ descriptions of fields we know how to
+                                   --   parse
+            -> UnrecFieldParser a  -- ^ possibly do something with
+                                   --   unrecognized fields
+            -> a                   -- ^ accumulator
+            -> [Field Position]    -- ^ fields to be parsed
+            -> ParseResult a
+parseFields descrs unrec ini fields = return ini -- TODO!
+
+-- TODO: build-depends is a special case
+-- TODO: add post check
+parseCondTree
+    :: forall a c.
+       [FieldDescr a]      -- ^ Field descriptions
+    -> UnrecFieldParser a  -- ^ How to parse unknown fields
+    -> (a -> c)            -- ^ Condition extractor
+    -> a                   -- ^ Initial value
+    -> [Field Position]    -- ^ Fields to parse
+    -> ParseResult (CondTree ConfVar c a)
+parseCondTree descs unread cond ini fields = do
+    x <- foldM go ini fields
+    return $ CondNode x (cond x) mempty
+  where
+    -- | TODO: change to take and return condnode
+    go :: a -> Field Position -> ParseResult a
+    go _ (Section (Name pos name) _ _) =
+        parseFailure pos $ "invalid subsection " ++ show name
+    go x (IfElseBlock pos _ _ _) = do
+        -- TODO
+        parseWarning pos "cannot parse if-else-blocks yet"
+        return x
+    go x (Field (Name pos name) fieldLines) =
+        case Map.lookup name fieldParsers of
+            Nothing -> do
+                parseWarning pos $ "Unknown field: " ++ show name
+                return x
+            Just parser ->
+                runFieldParser (parser x) fieldLines
+
+    fieldParsers :: Map FieldName (a -> FieldParser a)
```

Comment:
I'd love it if there was an explanation somewhere of why it would be really painful to do the parsers in a non-accumulating style. I think the reason is because you don't know a priori what all the fields are going to be, so a parser which returns just the value (rather than field: value pair) is just not that useful.